lib/backend/atomicwrite.go

```diff
@@ -0,0 +1,227 @@
+	KindNop ActionKind = 1 + iota
+
+	// KindPut indicates that the associated item should be written to the target key.
+	KindPut
+
+	// KindDelete indicates that any item at the target key should be removed.
+	KindDelete
+)
+
+// Action specifies an action to be taken against a backend item.
+type Action struct {
+	// Kind is the kind of action represented.
+	Kind ActionKind
+
+	// Item is the item to be written (only used when Kind is Put).
+	Item Item
+}
+
+// IsZero checks if this action appears unspecified.
+func (a *Action) IsZero() bool {
+	return a.Kind == 0
+}
+
+// Nop builds an action that does nothing.
+func Nop() Action {
+	return Action{
+		Kind: KindNop,
+	}
+}
+
+// Put builds an action that writes the provided item to the target key.
+func Put(item Item) Action {
+	return Action{
+		Kind: KindPut,
+		Item: item,
+	}
+}
+
+// Delete builds an action that removes the target key.
+func Delete() Action {
+	return Action{
+		Kind: KindDelete,
+	}
+}
+
+// ConditionalAction specifies a condition and an action associated with a given key. The condition
+// must hold for the action to be taken.
+type ConditionalAction struct {
+	// Key is the key against which the associated condition and action are to
+	// be applied.
+	Key []byte
+
+	// Condition must be one of Exists|NotExists|Revision(<revision>)|Whatever
+	Condition Condition
+
+	// Action must be one of Put(<item>)|Delete|Nop
+	Action Action
+}
+
+// Check validates the basic correctness of the conditional action.
+func (c *ConditionalAction) Check() error {
+	if len(c.Key) == 0 {
+		return trace.BadParameter("conditional action missing required parameter 'key'")
+	}
+
+	if c.Condition.IsZero() {
+		return trace.BadParameter("conditional action for %q missing required parameter 'condition'", c.Key)
+	}
+
+	if c.Action.IsZero() {
+		return trace.BadParameter("conditional action for %q missing required parameter 'action'", c.Key)
+	}
+
+	switch c.Condition.Kind {
+	case KindWhatever, KindExists, KindNotExists, KindRevision:
+	default:
+		return trace.BadParameter("conditional action for %q contains unexpected condition kind %v", c.Key, c.Condition.Kind)
+	}
+
+	switch c.Action.Kind {
+	case KindNop, KindDelete:
+	case KindPut:
+		if len(c.Action.Item.Value) == 0 {
+			return trace.BadParameter("conditional action for %q missing required put parameter Item.Value", c.Key)
+		}
+	default:
+		return trace.BadParameter("conditional action for %q contains unexpected action kind %v", c.Key, c.Action.Kind)
+	}
+
+	return nil
+}
+
+const (
+	// MaxAtomicWriteSize is the maximum number of conditional actions that may
+	// be applied via a single atomic write. The exact number is subject to change
+	// but must always be less than the minimum value supported across all backends.
+	MaxAtomicWriteSize = 64
+)
+
+// AtomicWrite is a standalone interface for the AtomicWrite method. This interface will be deprecated
+// once all backends implement AtomicWrite.
+type AtomicWrite interface {
```

Comment:
```suggestion
type AtomicWriter interface {
```
 
Similar to io.Writer exposing a Write method.